slidechain/watch.go

```diff
@@ -117,56 +117,97 @@ func (c *Custodian) watchExports(ctx context.Context) {
+
+			// Record the export in the db,
+			// then wake up a goroutine that executes peg-outs on the main chain.
+			const q = `
+				INSERT INTO exports 
+				(txid, exporter, amount, asset_xdr, temp_addr, seqnum, anchor, pubkey)
+				VALUES ($1, $2, $3, $4, $5, $6, $7, $8)`
+			_, err = c.DB.ExecContext(ctx, q, tx.ID.Bytes(), info.Exporter, info.Amount, info.AssetXDR, info.TempAddr, info.Seqnum, info.Anchor, info.Pubkey)
+			if err != nil {
+				log.Fatalf("recording export tx: %s", err)
+			}
 
-				// Check this Stellar asset code corresponds to retiredAssetIDBytes.
-				gotAssetID32 := txvm.AssetID(importIssuanceSeed[:], info.AssetXDR)
-				if !bytes.Equal(gotAssetID32[:], retiredAssetIDBytes) {
-					continue
-				}
+			log.Printf("recorded export: %d of txvm asset %x (Stellar %x) for %s", info.Amount, exportedAssetBytes, info.AssetXDR, info.Exporter)
 
-				var exporter xdr.AccountId
-				err = exporter.SetAddress(info.Exporter)
+			c.exports.Broadcast()
+		}
+	}
+}
+
+// Runs as a goroutine.
+func (c *Custodian) watchPegOuts(ctx context.Context, pegouts <-chan pegOut) {
+	defer log.Print("watchPegOuts exiting")
+
+	ticker := time.NewTicker(time.Minute)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			const q = `SELECT amount, asset_xdr, exporter, temp_addr, seqnum, anchor, pubkey FROM exports WHERE pegged_out IN ($1, $2)`
+			var (
+				txids, anchors, assetXDRs, pubkeys [][]byte
+				amounts, seqnums                   []int64
+				exporters, tempAddrs               []string
+				peggedOuts                         []pegOutState
+			)
+			err := sqlutil.ForQueryRows(ctx, c.DB, q, pegOutOK, pegOutFail, func(txid []byte, amount int64, assetXDR []byte, exporter, tempAddr string, seqnum, peggedOut int64, anchor, pubkey []byte) {
+				txids = append(txids, txid)
+				amounts = append(amounts, amount)
+				assetXDRs = append(assetXDRs, assetXDR)
+				exporters = append(exporters, exporter)
+				tempAddrs = append(tempAddrs, tempAddr)
+				seqnums = append(seqnums, seqnum)
+				peggedOuts = append(peggedOuts, pegOutState(peggedOut))
+				anchors = append(anchors, anchor)
+				pubkeys = append(pubkeys, pubkey)
+			})
+			if err != nil {
+				log.Fatalf("querying peg-outs: %s", err)
+			}
+			for i, txid := range txids {
+				err = c.doPostPegOut(ctx, assetXDRs[i], anchors[i], txid, amounts[i], seqnums[i], peggedOuts[i], exporters[i], tempAddrs[i], pubkeys[i])
 				if err != nil {
-					continue
+					log.Fatalf("doing post-peg-out: %s", err)
 				}
-
-				// Record the export in the db,
-				// then wake up a goroutine that executes peg-outs on the main chain.
-				const q = `
-					INSERT INTO exports 
-					(txid, exporter, amount, asset_xdr, temp_addr, seqnum)
-					VALUES ($1, $2, $3, $4, $5, $6)`
-				_, err = c.DB.ExecContext(ctx, q, tx.ID.Bytes(), exporter.Address(), retiredAmount, info.AssetXDR, info.TempAddr, info.Seqnum)
+			}
+		case p, ok := <-pegouts:
+			if ok {
```

Comment:
Change this to `if !ok { log.Fatalf(...) }` and then don't indent the normal-flow logic. (This is the [indent error flow](https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow) style guideline we've discussed before.)